Clean up in the value set API

romainbrenguier opened the pull request:
Several methods/members of value set are actually not used and should be deprecated to simplify the API. 
The `get_value_set` interface in which the result has to be given as argument is also not clear and it is better to return the result.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [na] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

romainbrenguier committed:
get_value_set takes expr by copy

This avoids having to create a tmp expression and allows moving into the function.

romainbrenguier committed:
Deprecate offset_is_zero

This is unused

romainbrenguier committed:
Deprecate value_sett::get

This is not implemented.

romainbrenguier committed:
Deprecate dynamic_object_id_sett

This is unused.

romainbrenguier committed:
Add versions of get_value_set returning a value

These return the computed value, which makes the information flow
clearer in the program.
We deprecate at the same time the functions from the previous version.

tautschnig commented:
@romainbrenguier Various CI steps in Travis are failing.

tautschnig commented on src/pointer-analysis/value_set.h:
```diff
@@ -90,70 +90,15 @@ class value_sett
-  /// to the internal map.
+  /// instead of a set of pairs to make lookup by `exprt` easier.
   class object_map_dt
+    : public std::map<object_numberingt::number_type, offsett>
```
This basically is a revert of 3b5263ff. We should have really good reasons if that's what should happen.

romainbrenguier replied:
Ok I see. I will drop that commit for now and try to find a better solution

tautschnig commented on src/pointer-analysis/value_set.h:
```diff
@@ -276,6 +277,7 @@ class value_sett
 
   /// Appears to be unimplemented.
+  DEPRECATED(SINCE(2019, 05, 22, "Unimplemented"))
   expr_sett &get(const irep_idt &identifier, const std::string &suffix);
```
I think we can just remove what isn't even implemented.

romainbrenguier replied:
I don't know. In test-gen there were cases of cpp files used instead of the cbmc ones which could add new implementation not existing in cbmc

smowton commented on src/pointer-analysis/goto_program_dereference.h:
```diff
@@ -65,9 +65,12 @@ class goto_program_dereferencet:protected dereference_callbackt
   void
   get_value_set(const exprt &expr, value_setst::valuest &dest) const override;
 
+  std::list<exprt> get_value_set(const exprt &expr) const override;
```
If we're adding a new entry-point we probably shouldn't use `std::list`, there's no call for a doubly-linked list here AFAICT

romainbrenguier replied:
but it is a list, and the fact that `value_setst::valuest` is a list is public, by writing `value_setst::valuest` we are just making it more complicated for the reader

smowton replied:
I'm not suggesting `valuest`, I'm suggesting `std::vector` because that's the most performant C++ sequence container when you don't have a good reason for something else (e.g. deque or list)

romainbrenguier replied:
oh right

smowton commented:
@owen-jones-diffblue could you check if `local_value_sett` uses `location_number`?

smowton reviewed (changes requested):
Suggest dropping the `std::map`-inheriting commit, which is a straight revert of https://github.com/diffblue/cbmc/pull/919, for the reasons given in that PR

romainbrenguier commented:
@tautschnig @smowton here is a PR to go instead of the std::map inheritance https://github.com/diffblue/cbmc/pull/4694

owen-mc-diffblue commented:
@smowton It does, but we can just add that field ourselves

romainbrenguier commented:
actually it seems location_number is used by value_set_domain_templatet so I will have to drop that commit too...

romainbrenguier committed:
Use vector instead of list for get_value_set result

vectors are generally more efficient than list, unless we use
concatenation or insertion which doesn't appear to be case here.

romainbrenguier commented:
@smowton I have now replace the use of list by vector

smowton commented on src/goto-symex/symex_dereference_state.h:
```diff
@@ -34,9 +34,12 @@ class symex_dereference_statet:
   goto_symext::statet &state;
   const namespacet &ns;
 
+  DEPRECATED(SINCE(2019, 05, 22, "use list returning version instead"))
```
vector returning

smowton commented on src/pointer-analysis/dereference_callback.h:
```diff
@@ -29,9 +29,12 @@ class dereference_callbackt
 public:
   virtual ~dereference_callbackt() = default;
 
+  DEPRECATED(SINCE(2019, 05, 22, "use list returning version instead"))
```
vector returning

smowton commented on src/pointer-analysis/value_sets.h:
```diff
@@ -28,12 +28,19 @@ class value_setst
   typedef std::list<exprt> valuest;
 
   // this is not const to allow a lazy evaluation
+  DEPRECATED(SINCE(2019, 05, 22, "use list returning version instead"))
```
vector returning

romainbrenguier committed:
Predefine DEPRECATED macro in doxyfile

This prevents the macro from making doxyfile miss method definitions
that are deprecated and automatically add the deprecated methods to the
Deprecated List generated by doxygen

romainbrenguier commented:
@owen-jones-diffblue @smowton I've added a commit which gets rid of doxygen warning that were generated for  this pull request, in case you want to have a look

tautschnig commented:
@romainbrenguier Please rebase now that #4729 is in to make CI happy.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 27f87e83a40685b0073ef88e1c34dcb65f9c2e8e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/113842424

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: fa1412c758edd7d4f50c92ddc73fbf5ba34791ae).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/114073541
